qutip/core/data/convert.pyx

```
@@ -259,63 +261,171 @@ cdef class _to:
+                The priority associated with this creator. Higher priority
+                conditions will be tested first and the first valid creator
+                (i.e. for which ``condition(object) == True``) will handle
+                the creation.
+
+        Notes
+        -----
+            Default creators are added with the following priorities:
+
+                * Objects that are instances of data-layer types are
+                  converted using ``.copy`` with priority ``100``.
+                * Objects that are ``numpy.ndarray`` instances or for
+                  which ``scipy.sparse.isspmatrix_csr`` is ``True`` are
+                  converted using an internal CSR converter with
+                  priority ``80``.
+                * Objects for which ``scipy.sparse.issparse`` is ``True``
+                  are converted using an internal CSR converter with
+                  priority ``10``.
+        """
+        _creators = []
+        for creator in creators:
+            if len(creator) == 2:
+                condition, create = creator
+                priority = 10
+            elif len(creator) == 3:
+                condition, create, priority = creator
+            if not callable(condition):
+                raise TypeError(str(condition) + " is not a callable")
+            if not callable(create):
+                raise TypeError(str(create) + " is not a callable")
+            _creators.append((condition, create, priority))
+        self._creators = sorted(_creators,
+                                key=lambda creator: creator[2],
+                                reverse=True)
+
+    def __call__(self, arg, shape=None):
+        for condition, create, _ in self._creators:
+            if condition(arg):
+                return create(arg, shape)
+        # Fallback on numpy if we don't have any matching creator method
+        return _data.Dense(np.array(arg, dtype=np.complex128),
```

I wonder if we should rather add this as a default creator below and just raise an error here if no creator is found? That way the default creator can be overridden by setting something with higher priority.

---

Do we want to make `Qobj` from anything that can be made an array?
In v4, we never directly raised any error: for unknown type it fell-back to `scipy.csr_matrix` and let it fail.
Here I do the same, but with numpy instead of scipy.

This here should not be the default but the fallback, so any creator will overwrite it.

---

As it stands, I don't think it matters a great deal whether this is a creator with predicate `lambda arg: True` and priority `-inf`, or written in like this.  That would change if we had a method to _remove_ creators as well, but I can't think of any neat way to do that at all (we probably _really_ don't want to make the backing list public), so I don't really see a difference.

I think we probably do want to accept anything array-like - it's very convenient (and I'm pretty sure used in the test base) to be able to do `Qobj([[1, 0], [0, 1]])` and such like.